File: src/main/java/com/linrol/cn/tool/model/GitCmd.java
package com.linrol.cn.tool.model;

import com.intellij.dvcs.repo.Repository;
import com.intellij.openapi.diagnostic.Logger;
import com.intellij.openapi.project.Project;
import com.intellij.openapi.ui.Messages;
import com.intellij.openapi.vfs.VirtualFile;
import com.intellij.openapi.wm.ToolWindow;
import com.intellij.openapi.wm.ToolWindowManager;
import git4idea.commands.Git;
import git4idea.commands.GitCommand;
import git4idea.commands.GitCommandResult;
import git4idea.commands.GitLineHandler;
import com.linrol.cn.tool.toolwindow.ToolWindowConsole;

import git4idea.repo.GitRepository;
import java.awt.EventQueue;
import java.util.Arrays;
import java.util.List;

import static com.linrol.cn.tool.utils.StringUtils.isBlank;

public class GitCmd {

    private static final Logger logger = Logger.getInstance(GitCmd.class);

    Project project;

    GitRepository repository;

    VirtualFile root;

    GitLineHandler handler;

    public GitCmd(Project project, GitRepository repository) {
        this.repository = repository;
        this.project = project;
        this.root = repository.getRoot();
    }

    public Project getProject() {
        return project;
    }

    public void setProject(Project project) {
        this.project = project;
    }

    public Repository getRepository() {
        return repository;
    }

    public void setRepository(GitRepository repository) {
        this.repository = repository;
    }

    public GitLineHandler getHandler() {
        return handler;
    }

    public void setHandler(GitLineHandler handler) {
        this.handler = handler;
    }

    public VirtualFile getRoot() {
        return root;
    }

    public void setRoot(VirtualFile root) {
        this.root = root;
    }

    public GitCmd addParameters(String... parameters) {
        Arrays.asList(parameters).forEach(handler::addParameters);
        return this;
    }

    public GitCmd build(GitCommand command, String... parameters) {
        return build(command, Arrays.asList(parameters));
    }

    public GitCmd build(GitCommand command, List<String> parameters) {
        GitLineHandler handler = new GitLineHandler(getProject(), getRoot(), command);
        parameters.forEach(handler::addParameters);
        this.handler = handler;
        return this;
    }

    public GitCmd config(boolean silent, boolean stdoutSuppressed, String... urls) {
        this.handler.setUrls(Arrays.asList(urls));
        this.handler.setSilent(silent);
        this.handler.setStdoutSuppressed(stdoutSuppressed);
        return this;
    }

    public GitCommandResult run() {
        String runString = this.handler.printableCommandLine();
        logger.debug(runString);
        GitCommandResult ret = Git.getInstance().runCommand(this.handler);
        if (!ret.success()) {
            String errorString = ret.getErrorOutputAsJoinedString();
            logger.info(String.format("Git run command:%s failed case by:%s", runString, errorString));
            throw new RuntimeException(errorString);
        }
        return ret;
    }

    public static void clear() {
        ToolWindowConsole.clear();
    }

    public void log(String msg) {
        log(project, msg);
    }

    public static void log(Project project, String msg) {
        ToolWindow toolWindow = ToolWindowManager.getInstance(project).getToolWindow("77tool");
        if (toolWindow == null) {
            return;
        }
        try {
            toolWindow.activate(() -> {
                ToolWindowConsole.show();
                ToolWindowConsole.log(project, msg);
            });
            // EventQueue.invokeAndWait(() -> );
        } catch (Exception e) {
            e.printStackTrace();
        }
        // toolWindow.show();
    }

    public String showInputDialog() {
        try {
            StringBuffer input = new StringBuffer();
            EventQueue.invokeAndWait(() -> {
                String cm = Messages.showInputDialog(project, "Input commit message:", String.format("%s create branch merge request", project.getName()), Messages.getInformationIcon());
                if (!isBlank(cm)) {
                    if (!"null".equals(cm)) {
                        input.append(cm);
                    }
                }
            });
            return input.toString();
        } catch (Exception e){
            e.printStackTrace();
            return null;
        }
    }

    public String getRemoteUrl() {
        return repository.getRemotes().stream().flatMap(m -> m.getPushUrls().stream()).filter(f -> !isBlank(f)).findAny().orElse(null);
    }

    public String getCurrentBranchName() {
        return repository.getCurrentBranchName();
    }
}

File: src/main/java/org/intellij/tool/model/GitCmd.kt
package org.intellij.tool.model

import com.intellij.openapi.application.ApplicationManager
import com.intellij.openapi.diagnostic.logger
import com.intellij.openapi.progress.ProcessCanceledException
import com.intellij.openapi.progress.ProgressIndicator
import com.intellij.openapi.progress.ProgressManager
import com.intellij.openapi.progress.Task
import com.intellij.openapi.project.Project
import com.intellij.openapi.vcs.VcsException
import com.intellij.openapi.vfs.VirtualFile
import com.intellij.openapi.wm.ToolWindowManager
import git4idea.commands.Git
import git4idea.commands.GitCommand
import git4idea.commands.GitCommandResult
import git4idea.commands.GitLineHandler
import git4idea.repo.GitRepository
import org.apache.commons.lang3.exception.ExceptionUtils
import org.intellij.tool.toolwindow.ToolWindowConsole


class GitCmd(private var project: Project, private var repository: GitRepository) {
    var root: VirtualFile = repository.root

    private lateinit var handler: GitLineHandler

    private lateinit var command: GitCommand

    fun addParameters(vararg parameters: String): GitCmd {
        parameters.forEach { handler.addParameters(it) }
        return this
    }

    fun build(command: GitCommand, vararg parameters: String): GitCmd {
        this.command = command
        return build(command, parameters.toList())
    }

    private fun build(command: GitCommand, parameters: List<String>): GitCmd {
        this.command = command
        val handler = GitLineHandler(project, this.root, command)
        parameters.forEach { handler.addParameters(it) }
        this.handler = handler
        return this
    }

    fun config(silent: Boolean, stdoutSuppressed: Boolean, vararg urls: String): GitCmd {
        handler.urls = urls.toList()
        handler.setSilent(silent)
        handler.setStdoutSuppressed(stdoutSuppressed)
        return this
    }

    fun run(): GitCommandResult {
        try {
            val runString = handler.printableCommandLine()
            logger.debug(runString)
            val title = "Git ${this.command} running"
            val ret = ProgressManager.getInstance().run(object : Task.WithResult<GitCommandResult, VcsException>(project, title, true) {
                override fun compute(indicator: ProgressIndicator): GitCommandResult {
                    return Git.getInstance().runCommand(handler)
                }
            })
            // GitCommandResult ret = Git.getInstance().runCommand(this.handler);
            if (!ret.success()) {
                val errorString = ret.errorOutputAsJoinedString
                logger.info("Git run command:${runString} failed case by:${errorString}")
                throw RuntimeException(errorString)
            }
            return ret
        } catch (e: VcsException) {
            throw RuntimeException(ExceptionUtils.getRootCauseMessage(e))
        } catch (e: ProcessCanceledException) {
            throw RuntimeException(ExceptionUtils.getRootCauseMessage(e))
        }
    }

    fun log(msg: String) {
        log(project, msg)
    }

    val remoteUrl: String get() = repository.remotes.flatMap { it.pushUrls }.first { it.isNotBlank() }

    val currentBranchName: String get() = repository.currentBranchName.toString()

    companion object {
        private val logger = logger<GitCmd>()

        fun clear() {
            ToolWindowConsole.clear()
        }

        @JvmStatic
        fun log(project: Project, msg: String) {
            val toolWindow = ToolWindowManager.getInstance(project).getToolWindow("77tool") ?: return
            try {
                ApplicationManager.getApplication().invokeLater {
                    toolWindow.activate {
                        ToolWindowConsole.show()
                        ToolWindowConsole.log(project, msg)
                    }
                }
                // EventQueue.invokeAndWait(() -> );
            } catch (e: Exception) {
                e.printStackTrace()
            }
            // toolWindow.show();
        }
    }
}
